### DISCUSSION TITLE
Using x-cloak on parent div

### DISCUSSION THREAD

**Discussion by haztakki** [alpinejs/alpine#645] [4. General]
Hello all,
I have a modal inside a **fixed** element which by default is hidden and later shown when a button is clicked. I use **x-cloak** to hide it initially however I must apply it to the actual modal div and not the parent. I'm just wondering why is this? Say if I have a bunch of elements inside a **fixed** container then I'd have to apply it to each one rather than just the parent. Is there a reason why the parent can't control this? I assume this happens only on **absolute** and **fixed** elements? I haven't ran into a case where I need this but it was just a thought.

**Comment by ryangjchandler** [comment]
@HazJ have you got a code snippet that we can look at?

**Comment by haztakki** [comment]
I don't think it is a bug, I'm sure it is intended behaviour? I was just wondering about the reasoning that's all, anyway... I have a modal (from **TailwindUI**). The main modal **div** is inside the suggest-edit Livewire component. Note: This happens regardless of Livewire, if I apply **x-cloak only** on the parent **div** (the one you see below) then it doesn't cloak the child elements. It seems that I have to be specific. Is there a reason why **x-cloak** can't function on the parent and affect child elements? Also, is this related to elements that are using **absolute** and **fixed** positioning? I haven't tested it yet but this is my initial thought.
```html
        <div x-show="modal" class="fixed bottom-0 inset-x-0 px-4 pb-4 sm:inset-0 sm:flex sm:items-center sm:justify-center">
            <div class="fixed inset-0 transition-opacity">
                <div class="absolute inset-0 opacity-75"></div>
            </div>
            <livewire:suggest-edit/>
        </div>
```

```html
<div x-cloak> <!-- This does nothing to the child elements, they will still show, I have to apply it on the div below -->
    <div x-cloak></div>
</div>
```

**Reply by SimoTod** [reply]
I think there's a bit of confusion, `x-cloak` is an html attribute which, by itself, doesn't do anything. 
It gets removed once Alpine kicks in, that's all.
What hides the element is the css rule that you need to add to your stylesheet. If you follow the instructions, it's about hiding the element but, for example, you could change that css rule to
`[x-cloak] {background: red}` and the element will have a red background for a really short time and then become normal once alpine starts.

Normal css rules apply, Alpine doesn't implement any additional logic: if you hide the parent, it will also hide the children but, here's the catch, javascript is single thread so only one directive is evaluated at once and Alpine goes top to bottom. It means that it will hit the parent element first and remove the x-cloak which will temporary show the children. It will then shortly hit the child, evaluate the x-show and hide it again.
This may or may not cause some flashing depending on what your browser decides to do so it's better to put x-cloak on the elements which need it.

**Reply by haztakki** [reply]
`here's the catch, javascript is single thread so only one directive is evaluated at once and Alpine goes top to bottom. It means that it will hit the parent element first and remove the x-cloak which will temporary show the children. It will then shortly hit the child, evaluate the x-show and hide it again.
This may or may not cause some flashing depending on what your browser decides to do so it's better to put x-cloak on the elements which need it.`
Yeah, I guess that's why. Thanks - Can't seem to mark your comment as the answer. It'd be cool if the attribute affected the child elements at the same time (no flicker) just as the parent does.

**Reply by ryangjchandler** [reply]
As Simone said, Alpine does one thing at a time. You need the CSS rule since Alpine doesn't provide it and I'd always recommend putting x-cloak on the very end.

**Reply by haztakki** [reply]
Yeah, I already did, I had it working before my post. I guess my initial understanding was that "if it cloaks the parent, why do the child elements the flicker" but as explained, it doesn't work like that. Thanks.

**Reply by SimoTod** [reply]
You can add display none to the parent div and in x-init you can return a callback that gets the div and shows it. basically you manually implements x-cloak but, since you use a callback in x-init, it will run after the component has started and x-show has already been evaluated

**Reply by SimoTod** [reply]
P.s. I think it would be reasonable if it worked as you described at the beginning and x-cloak was removed only after the whole component had been initialised so feel free to send a PR if you want to.

**Comment by haztakki** [comment]
Hello SimoTod,
I understand what it does, I'm using it perfectly fine. It works. What I am getting at is, if I wanted to cloak a bunch of elements inside a **div** then I have to apply the attribute to each one. I would of thought that the parent **div** with the **x-cloak** attribute would of affected its child elements too. I was just wondering why it didn't. Thanks for explaining.
